Build on CBL-mariner host with rootfs

sbomer opened the pull request:
This changes the linux builds to use the CBL-mariner build images added in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/832. All of these builds are now cross-builds with a rootfs (including x64, and x64 musl).

- The new images intentionally don't have binutils, so our build jobs have been updated to use llvm-based tools. This includes lld when we run nativeaot over our tests. This leads to a slight size regression, I believe due to ILCompiler not supporting `--gc-sections` with lld (see https://github.com/dotnet/runtime/pull/83558#issuecomment-1473693469). cc @MichalStrehovsky
- This change turns off PGO optimization because it was hitting a compiler crash when consuming and old version of profile data produced by clang9 instrumented builds (https://bugzilla.redhat.com/show_bug.cgi?id=1827282). The intention is to quickly flow the change into our optimization repo, then collect and ingest the new data so we can turn optimization back on.
- The Mono LLVMAot test build is kept on the old build images, because they run the aot compiler which depends on binutils. We can probably move it to the same plan as the other legs, but I'm not familiar with the context here. cc @akoeplinger. We would at least need to fix `AS_NAME` here: https://github.com/dotnet/runtime/blob/1d2cd206f40306cbf06f3e2676e7cfa301077285/src/mono/mono/mini/aot-compiler.c#L13030
- Mono arm64 product build is kept on the old images due to https://github.com/dotnet/runtime/issues/84503.

Contributes to https://github.com/dotnet/runtime/issues/83428.

janvorli committed:
Enable x64 Alpine crossbuild

This change enables cross-building for Alpine x64 on other x64 distro.

sbomer committed:
Set PKG_CONFIG_SYSROOT_DIR

Necessary because pkg-config on mariner is a shim for
pkgconf, which doesn't support override variables.

sbomer committed:
More build fixes

- Use lld for LTO instead of ld.gold
- Set cross toolchain options for x64 mono build
- Prefer llvm-objcopy for mono symbol stripping

sbomer committed:
test

sbomer committed:
Print full clang command to show selected linker

sbomer committed:
Build scripts

sbomer committed:
Fix x64 -> x64 musl cross-build

sbomer committed:
Enable thumb-2 in mono build

Our custom LLVM build seems to have different defaults when targeting
ARM than the clang-12 available in ubuntu.
Probably related to
https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/commit/b7363248b115339c4fb838fcd3ae43671eedae0a.

Passing the -march explicitly should not hurt.

sbomer committed:
Allow cross compiling for x86

sbomer committed:
Enable x86 build with lld

Allow text relocations. The lld defaults differ from ld, which will add
text relocations on demand by default.

See https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities

sbomer committed:
Fix x86 mono cross-build

sbomer committed:
Some cleanup

sbomer committed:
Add gcc-toolchain arg when aot-ing crossgen

+ more cleanup

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerTest

sbomer committed:
Fix AOT compilation of ILCompiler and crossgen2

sbomer committed:
Only pass --gcc-toolchain where required

To locate gcc toolchain in target sysroot, when host arch matches target arch.

sbomer committed:
Cleanup

sbomer committed:
Fix symbol stripping on ci builds

sbomer committed:
Fix nativeaot test build

sbomer committed:
Pick up mariner images, set cross-build arguments

am11 commented on eng/Subsets.props:
```diff
@@ -249,8 +249,8 @@
-  <!-- Build the CoreCLR cross-arch tools when we're doing a cross-architecture build and either we're building any CoreCLR native tools for platforms CoreCLR fully supports or when someone explicitly requests them -->
-  <ItemGroup Condition="(('$(ClrRuntimeBuildSubsets)' != '' and '$(PrimaryRuntimeFlavor)' == 'CoreCLR' and '$(TargetsMobile)' != 'true') or $(_subset.Contains('+clr.crossarchtools+'))) and '$(BuildArchitecture)' != '$(TargetArchitecture)'">
+  <!-- Build the CoreCLR cross tools when we're doing a cross build and either we're building any CoreCLR native tools for platforms CoreCLR fully supports or when someone explicitly requests them -->
+  <ItemGroup Condition="(('$(ClrRuntimeBuildSubsets)' != '' and '$(PrimaryRuntimeFlavor)' == 'CoreCLR' and '$(TargetsMobile)' != 'true') or $(_subset.Contains('+clr.crossarchtools+'))) and '$(CrossBuild)' == 'true'">
```
This is not very obvious but `CrossBuild` property (or `build -cross`) only covers targets which are using `toolchain.cmake` file during cross-compilation. Other targets, such as macOS, Windows and Android, do not set this property during the cross build (e.g. target: osx-arm64, host: osx-x64 bypasses `CrossBuild` and its effect on various subsets build). Therefore, the previous condition is more accurate as we want to build crossarchtools on those targets as well.

am11 replied:
Currently, I think this is intentional. e.g. `freebsd-x64` and `illumos-x64` builds also use the same prereq images crossdeps layer and run build on `linux-x64`. So we are skipping this build on those platforms.

One of these two options will work:

```diff
- .. and '$(BuildArchitecture)' != '$(TargetArchitecture)'`
+ .. and ('$(BuildArchitecture)' != '$(TargetArchitecture)' or '$(CrossBuild)' != 'true')
```
(preferred; it will not the most scalable solution but it will work if we don't break the `build-rootfs` contract 🙂)

or 
```diff
- .. and '$(BuildArchitecture)' != '$(TargetArchitecture)'`
+ .. and ('$(BuildArchitecture)' != '$(TargetArchitecture)' or '$(HostOS)' != '$(TargetOS)')
```
(tricky because mobile-like targets - and possibly some others - always have a different [`host` and `build`](https://stackoverflow.com/a/15901574/863980) but we don't make this distinction clearly outside `src/mono`)

am11 commented on eng/common/cross/build-rootfs.sh:
Just a [reminder](https://github.com/dotnet/arcade/blob/main/eng/common/README.md), please open a parallel PR upstream (in dotnet/arcade) for changes under `eng/common` dir. Otherwise, the next subsequent arcade auto-sync will undo these changes.

am11 commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -37,10 +39,12 @@ resources:
 
     - container: linux_x64
-      image: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7
+      image: mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-amd64
```
Great achievement! 🏅👍

am11 commented on src/coreclr/pgosupport.cmake:
```diff
@@ -6,7 +6,7 @@ if(NOT WIN32)
   function(check_have_lto)
     set(CMAKE_REQUIRED_FLAGS -flto)
-    set(CMAKE_REQUIRED_LIBRARIES -flto -fuse-ld=gold)
+    set(CMAKE_REQUIRED_LIBRARIES -flto -fuse-ld=lld)
```
We don't need to re-specify linker flavor here, it is inserted globally:
```suggestion
    set(CMAKE_REQUIRED_LIBRARIES -flto)
```
(unless there was a reason to keep `gold` linker here, in which case, undo)

sbomer committed:
Don't use lld on macos

sbomer committed:
Update src/coreclr/pgosupport.cmake

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

sbomer committed:
Don't pass clang9

sbomer committed:
Fix osx linker in tests too

sbomer committed:
Define separate platform for GCC build

This allows passing compilerName: gcc without also
passing crossBuild: true (which is the new default for
the linux_x64 platform).

sbomer committed:
Disable PGO for linux x64 and musl-x64

sbomer committed:
Build cross tools for x64 and musl-x64

sbomer committed:
Pass crossArg to mono build

sbomer committed:
Pass cross arg in build-test-job.yml

sbomer committed:
Fix typo

sbomer committed:
Pass through no-PGO arg in global-build-job.yml

sbomer committed:
Remove pgoType from job name

To fix dependencies between jobs

sbomer committed:
Pass noPgo arg to native prereqs step

Otherwase it sets a variable tha causes pgodatapath to be passed to next step

sbomer committed:
Fix crossArg not defined for NativeAOT test build

sbomer committed:
Pass crossBuild through build-test-job.yml

To ensure it reaches common pri0 test build

sbomer committed:
Set linker flavor, sysroot in nativeaot library tests

sbomer committed:
Use git diff for AsnXml

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerBuild

sbomer committed:
Pass crossBuild through mono's xplat-pipeline-job.yml

sbomer committed:
Move crossArg after runtime build params

Should fix Common Pri0 test build

sbomer committed:
Disable PGO from MSBuild instead

sbomer committed:
Fix empty container

sbomer committed:
Pass crossArg to mono Build packages step

sbomer committed:
Fix gcc container definition

sbomer committed:
Use init-compiler.sh for mono cross build

Also replace 'which' with 'command -v'

sbomer committed:
Condition linker selection based on compiler

To help with gcc builds. Same may need to be done for nativeaot test builds.

sbomer committed:
Fix publish condition

sbomer committed:
Pass compilerArg to managed component build

Should allow clr.tools to build, using gcc to AOT-compile ILCompiler

sbomer committed:
Use command -v instead of which

(Missing from earlier commit)

sbomer committed:
Use crossArg instead of _crossBuildPropertyArg

Should let nativeaot test native build use toolchain file.

sbomer committed:
Set mono toolchain for aot build

sbomer committed:
Adjust DwarfDump baseline, show expected AOT size

sbomer committed:
Don't use new images for Mono LLVMAot test build

sbomer committed:
Build cross tools when BuildArch != TargetArch

sbomer committed:
Fix typo

sbomer committed:
Don't set crossBuild for mono llvmaot

sbomer committed:
Fix ILCompiler, crossgen invocation for windows

sbomer committed:
Add PGO instrumentation legs

sbomer committed:
Run source-build legs

sbomer committed:
Pick up arm images

sbomer committed:
Fix indentation

sbomer committed:
Undo job name change

sbomer committed:
Add linux arm64 libraries build

sbomer committed:
Don't run AOT'd crossgen that targets musl

sbomer committed:
Fix crossgen-corelib logic

BuildDll should be true for x64 and x64-alpine builds

sbomer committed:
Freeze image versions

Until we can use LLVM 16

sbomer committed:
Use existing images for mono arm64 build

sbomer committed:
Don't trigger source-build

This is broken in rolling builds already

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerBuild

sbomer committed:
Update dwarfdump, size baseline

There are fewer dwarfdump warnings when building with lld.
We don't support --gc-sections when ILCompiler uses lld.

jkoritzinsky commented on eng/common/cross/build-rootfs.sh:
Make sure to open a PR in Arcade with the eng/common changes so they don't get reverted here in the next Arcade update after this is merged.

jkoritzinsky commented on eng/native/tryrun.cmake:
```diff
@@ -12,6 +12,7 @@ if(EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/armv7-alpine-linux-musleabihf OR
    EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/s390x-alpine-linux-musl OR
    EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/ppc64le-alpine-linux-musl OR
    EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/i586-alpine-linux-musl OR
+   EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/x86_64-alpine-linux-musl OR
```
Do we also need a `tryrun.cmake` configuration for the regular glibc x64 build? I'm worried we'll determine our results based on CBL-Mariner's runtime libraries, not the results the correspond to our "target" rootfs.

jkoritzinsky commented on eng/nativepgo.targets:
```diff
@@ -1,7 +1,8 @@
   <PropertyGroup>
     <NativeOptimizationDataSupported Condition="'$(TargetOS)' == 'windows' And ('$(TargetArchitecture)' == 'x64' Or '$(TargetArchitecture)' == 'x86' Or '$(TargetArchitecture)' == 'arm64')">true</NativeOptimizationDataSupported>
-    <NativeOptimizationDataSupported Condition="'$(TargetOS)' == 'linux' And ('$(TargetArchitecture)' == 'x64' Or '$(TargetArchitecture)' == 'arm64')">true</NativeOptimizationDataSupported>
+    <!-- Disable PGO until we get profile data collected with a new clang -->
+    <NativeOptimizationDataSupported Condition="'$(TargetOS)' == 'linux' And ('$(TargetArchitecture)' == 'x64' Or '$(TargetArchitecture)' == 'arm64')">false</NativeOptimizationDataSupported>
```
When we're ready to merge this, file a follow up issue to re-enable consuming profile data?

jkoritzinsky commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -5,7 +5,7 @@ parameters:
   containers:
     - container: linux_arm
-      image: mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-arm
+      image: mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-arm-0230407130525-290f679
```
Why are we using the versioned container IDs? Didn't we move to the versionless ones to ensure we'd always consume security updates automatically?

jkoritzinsky commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -14,21 +14,28 @@ resources:
       env:
         ROOTFS_DIR: /crossrootfs/armv6
 
+    - container: mono_linux_arm64
+      image: cr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04-cross-arm64
+      env:
+        ROOTFS_DIR: /crossrootfs/arm64
+
```
Why are we using an ubuntu base for this particular target. Is this the particular case in our build where we have to care about the differences between build, host, and target platforms as all three differ?

jkoritzinsky replied:
Can we put a mention of the issue number in the YAML?

jkoritzinsky commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -59,6 +68,12 @@ resources:
+    - container: debian-11-gcc12-amd64
+      image: mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-gcc12-amd64
+
+    - container: linux_x64_llvmaot
+      image: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7
+
```
Why are we keeping the LLVMAOT build on the centos7 image? Same as the previous case?

Keeping this on CentOS 7 will delay some future infrastructure improvements in the repo (like upgrading CMake).

akoeplinger replied:
hmm yeah we should be able to pass in the clang tools instead of binutils (I know xamarin-android does this already), can you please file a separate issue so we can take a look at that?

jkoritzinsky commented on eng/pipelines/runtime.yml:
```diff
@@ -165,6 +163,35 @@ extends:
                 eq(dependencies.evaluate_paths.outputs['SetPathVars_runtimetests.containsChange'], true),
                 eq(variables['isRollingBuild'], true))
 
+      # Temporarily define PGO jobs to get PR validation of PGO instrumented builds
+      - template: /eng/pipelines/common/platform-matrix.yml
+        parameters:
+          jobTemplate: /eng/pipelines/coreclr/templates/build-job.yml
+          buildConfig: release
+          platforms:
+          - linux_x64
+          - linux_arm64
+          jobParameters:
+            isOfficialBuild: ${{ variables.isOfficialBuild }}
+            signBinaries: false
+            testGroup: innerloop
+            pgoType: 'PGO'
+
+
+      # Temporarily define PGO jobs to get PR validation of PGO instrumented builds
+      - template: /eng/pipelines/common/platform-matrix.yml
+        parameters:
+          jobTemplate: /eng/pipelines/installer/jobs/build-job.yml
+          buildConfig: Release
+          jobParameters:
+            isOfficialBuild: ${{ variables.isOfficialBuild }}
+            liveRuntimeBuildConfig: release
+            liveLibrariesBuildConfig: Release
+            pgoType: 'PGO'
+          platforms:
+          - linux_x64
+          - linux_arm64
```
If this is temporary, make sure to either remove it before merging this PR or file an issue to track removing it at the correct time.

jkoritzinsky commented on eng/testing/tests.singlefile.targets:
```diff
@@ -24,11 +24,12 @@
   <PropertyGroup Condition="'$(TestNativeAot)' == 'true'">
     <IlcToolsPath>$(CoreCLRILCompilerDir)</IlcToolsPath>
     <IlcToolsPath Condition="'$(TargetArchitecture)' != '$(BuildArchitecture)'">$(CoreCLRCrossILCompilerDir)</IlcToolsPath>
```
This condition should be updated to ensure that our NativeAOT libraries tests link against the target rootfs instead of the host image binaries.

jkoritzinsky commented on src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:
We have very similar logic for determining the target triple as well as the `--gcc-toolchain` argument in the [NativeExports](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj) test project. Can we share some of this logic?

jkoritzinsky commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -57,11 +61,36 @@
       </ItemGroup>
   </Target>
 
+  <!-- Needed for the amd64 -> amd64 musl cross-build to pass the target flag. -->
+  <Target Name="_FixIlcTargetTriple"
+          AfterTargets="SetupOSSpecificProps"
+          Condition="'$(CrossBuild)' == 'true' and '$(HostOS)' != 'windows'">
+    <!-- Compute CrossCompileRid, and copy the downstream logic as-is. -->
+    <PropertyGroup>
+      <CrossCompileRid>$(RuntimeIdentifier)</CrossCompileRid>
+
+      <CrossCompileArch />
+      <CrossCompileArch Condition="$(CrossCompileRid.EndsWith('-x64'))">x86_64</CrossCompileArch>
+      <CrossCompileArch Condition="$(CrossCompileRid.EndsWith('-arm64')) and '$(_IsApplePlatform)' != 'true'">aarch64</CrossCompileArch>
+      <CrossCompileArch Condition="$(CrossCompileRid.EndsWith('-arm64')) and '$(_IsApplePlatform)' == 'true'">arm64</CrossCompileArch>
+
+      <TargetTriple />
+      <TargetTriple Condition="'$(CrossCompileArch)' != ''">$(CrossCompileArch)-linux-gnu</TargetTriple>
+      <TargetTriple Condition="'$(CrossCompileArch)' != '' and ($(CrossCompileRid.StartsWith('linux-musl')) or $(CrossCompileRid.StartsWith('alpine')))">$(CrossCompileArch)-alpine-linux-musl</TargetTriple>
+      <TargetTriple Condition="'$(CrossCompileArch)' != '' and ($(CrossCompileRid.StartsWith('freebsd')))">$(CrossCompileArch)-unknown-freebsd12</TargetTriple>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <LinkerArg Include="--target=$(TargetTriple)" Condition="'$(TargetOS)' != 'osx' and '$(TargetTriple)' != ''" />
+    </ItemGroup>
+  </Target>
+
```
Same logic sharing idea here as in `ILCompiler.csproj`.

jkoritzinsky replied:
Tbh I'm more interested in sharing the support from the NativeExports project with crossgen and ILC than sharing the infrastructure in this PR between these two projects. The infra in NativeExports uses our native CMake build components to find all of the same tooling and flags we use in the cross build for the DNNE build, which will be more reliable than any other attempt to construct the target triple (and it will ensure consistency between the builds).

jkoritzinsky commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -1,7 +1,7 @@
   <PropertyGroup>
     <AsnXml />
-    <_AsnXmlDiffCmd>%24%28command -v diff%29 -q -a</_AsnXmlDiffCmd>
+    <_AsnXmlDiffCmd>%24%28command -v git%29 diff --quiet</_AsnXmlDiffCmd>
```
This change is due to the difference in tools that's available on CBL-Mariner?

am11 replied:
We explicitly support building tarball (source build) without git in arcade. Installing diffutils makes sense here, IMO, since it is available via base packages on other distros.

jkoritzinsky replied:
The source-build build in dotnet/installer validates this behavior I think. In any case, RedHat and Caconical build the product in that mode (at least until the VMR comes online fully).

jkoritzinsky commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -124,7 +124,7 @@ endif()
 install (TARGETS System.Globalization.Native-Static DESTINATION ${STATIC_LIB_DESTINATION} COMPONENT libs)
 
-if(NOT CLR_CMAKE_TARGET_APPLE AND NOT CLR_CMAKE_TARGET_ANDROID)
+if(NOT CLR_CMAKE_TARGET_APPLE AND NOT CLR_CMAKE_TARGET_ANDROID AND NOT CLR_CMAKE_TARGET_ALPINE_LINUX)
```
It might be worth it to enhance the `verify-so.sh` script to be able to handle cross-build scenarios, though I don't know how easy that would be. This would be a good follow-up issue.

janvorli replied:
The targets excluded here don't need the verification. Linkers on these platforms don't succeed linking executables with undefined symbols.

jkoritzinsky commented on src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Baseline.csproj:
```diff
@@ -7,10 +7,7 @@
-    <!-- We should be able to delete once CI machines no longer use ancient LLVM -->
-    <ObjCopyName>objcopy</ObjCopyName>
-  </PropertyGroup>
+    </PropertyGroup>
```
```suggestion
  </PropertyGroup>
```

jkoritzinsky reviewed (changes requested):
Looks pretty good, but there's a few changes I'd like.

sbomer committed:
Fix yaml mistakes

sbomer committed:
Update src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Baseline.csproj

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

sbomer committed:
PR feedback

- Fix condition in tests.singlefile.targets
- Add issue link

Also actually remove linux_arm64 mono leg

sbomer committed:
Fix arm image tags

am11 commented on src/mono/mono.proj:
In https://github.com/dotnet/runtime/pull/83903, we have removed the hardcoded list of objcopy to use `configuretools.cmake` (which `coreclr,libs.native,host.native` use). If we merge that, the diffs here will shrink down a bit.

akoeplinger replied:
I left a few comments there, it looks pretty much ready so we could get that one in before.

am11 commented on src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Baseline.csproj:
```diff
@@ -7,9 +7,6 @@
     <StripSymbols>true</StripSymbols>
-    
-    <!-- We should be able to delete once CI machines no longer use ancient LLVM -->
-    <ObjCopyName>objcopy</ObjCopyName>
```
👍

am11 commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -75,7 +73,7 @@
 
   <Target Name="RunPublishedCrossgen" AfterTargets="PublishCrossgen"
-          Condition="'$(TargetOS)' == '$(HostOS)' and '$(TargetArchitecture)' == '$(BuildArchitecture)'">
+          Condition="'$(TargetOS)' == '$(HostOS)' and '$(TargetArchitecture)' == '$(BuildArchitecture)' and '$(TargetsLinuxMusl)' != 'true'">
```
```suggestion
          Condition="'$(TargetOS)' == '$(HostOS)' and '$(TargetArchitecture)' == '$(BuildArchitecture)' and '$(CrossBuild)' != 'true'">
```

This way if we are building on linux-musl-x64, this target will run and only skip if we are cross-building for linux-musl-x64 on linux-x64.

sbomer committed:
Update src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerBuild

sbomer committed:
Revert "Write linker script for lld to enable gc-sections (#84493)"

This reverts commit 3ab663e128e233ad1f4a1a4a0df64b08a846bcd5.

sbomer committed:
Update image tags

To pick up diffutils

sbomer committed:
Revert "Use git diff for AsnXml"

This reverts commit a51dbde024bf8fdb294a17a4239b0b342168de24.

sbomer committed:
Fix nativeaot libraries test build

sbomer committed:
Remove temporary PGO instrumentation legs

sbomer committed:
Fix musl_arm tag

sbomer committed:
Remove left-over leg

Was dependency for PGO instrumentation

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerBuild

sbomer committed:
Merge remote-tracking branch 'origin/main' into marinerBuild

janvorli commented on src/mono/mono.proj:
```diff
@@ -546,6 +560,7 @@
     <!-- Linux options -->
     <ItemGroup Condition="'$(TargetsLinux)' == true">
       <_MonoCFLAGS Include="-Wl,--build-id=sha1" />
+      <_MonoCFLAGS Condition="'$(Platform)' == 'arm'" Include="-march=armv7-a" />
```
I am not sure if mono should target armv7-a. I have thought it targeted armv6, but I am not sure. @akoeplinger do you know the truth here?

akoeplinger replied:
Platform=arm means armv7 in the dotnet/runtime build, we have a separate armv6 config which is community supported (and only works on Mono) so this looks good to me.

Can you check whether we passed -mfpu=vfpv3-d16 in the eisting ubuntu images?

akoeplinger replied:
You'll also want to add this to _MonoAOTCFLAGS

jkoritzinsky commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -59,6 +69,12 @@ resources:
       env:
         ROOTFS_DIR: /crossrootfs/ppc64le
 
+    - container: debian-11-gcc12-amd64
+      image: mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-gcc12-amd64
```
Not in this PR, but can we consider moving this image to be based off Ubuntu 22.04 instead of Debian 11 (Debian 11 has an older CMake version than the one we want to use and IIRC I don't think we want to use the Kitware feed in our builds)?

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -818,6 +839,13 @@
+    <ItemGroup Condition="'$(RealTargetArchitecture)' == 'x64' and '$(RealTargetOS)' == 'linux'">
+      <MonoAOTCMakeArgs Include="-DCMAKE_TOOLCHAIN_FILE=$(CrossToolchainFile)" />
+      <_MonoAotBuildEnv Include="TARGET_BUILD_ARCH=x64" />
+      <_MonAotoBuildEnv Include="PKG_CONFIG_PATH=$(MonoCrossDir)/usr/lib/x86_64-linux-gnu/pkgconfig" />
```
```suggestion
      <_MonoAotBuildEnv Include="PKG_CONFIG_PATH=$(MonoCrossDir)/usr/lib/x86_64-linux-gnu/pkgconfig" />
```

sbomer committed:
PR feedback

- Add armv7-a to _MonoAOTCFLAGS

Also similar for CXX flags

sbomer committed:
Update src/mono/mono.proj

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

sbomer committed:
Fix condition for mono aot flags

I think this should check RealTargetArchitecture, not Platform

janvorli reviewed (approved):
LGTM, thank you!

rzikm commented:
This PR (or a related one) breaks System.Net.Security.Enterprise.Test builds (https://dev.azure.com/dnceng-public/public/_build/results?buildId=243440&view=logs&j=03b95830-93c2-599b-7e4d-1efb940adc28&t=9abf97cd-b4a3-5a26-6cf2-d05ce865f2ef&l=3907).

```
clang : error : invalid linker name in argument '-fuse-ld=lld' [/repo/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
```

The docker image used (mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04) has clang-9 but no lld binary installed. Will that be fixed by the TODO at
https://github.com/sbomer/runtime/blob/92d1f2217649103e3f3e5121b8ad35dca5a81bbb/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj#L62
